### PR TITLE
Increase perf of `get_auth_chain_ids` used in state res v2.

### PR DESCRIPTION
**Pull request by erikjohnston**
We do this by moving the recursive query to be fully in the DB.

### CODE REVIEW

File: synapse/storage/data_stores/main/event_federation.py [hunk 1]
```diff
@@ -61,6 +62,28 @@ def get_auth_chain_ids(self, event_ids, include_given=False):
...
+            txn.execute(sql, (list(event_ids),))
+
+            results = set(event_id for event_id, in txn)
+
+            if include_given:
+                results.update(event_ids)
+
+            return list(results)
+
+        # Database doesn't necessarily support recursive CTE, so we falll
```
Comment (richvdh) [inline synapse/storage/data_stores/main/event_federation.py]: fall

File: synapse/storage/data_stores/main/event_federation.py [hunk 2]
```diff
@@ -61,6 +62,28 @@ def get_auth_chain_ids(self, event_ids, include_given=False):
         )
 
     def _get_auth_chain_ids_txn(self, txn, event_ids, include_given):
+        if isinstance(self.database_engine, PostgresEngine):
+            # For efficieny we make the database do this if we can.
```
Comment (richvdh) [inline synapse/storage/data_stores/main/event_federation.py]:
```suggestion
            # For efficiency we make the database do this if we can.
```

**Comment by erikjohnston** [comment]
> honestly I'll take your word for it on the postgres magic. lgtm other than typos

I got it wrong the first time round and everything exploded in a ball of flames, so hopefully its fine now :D

File: synapse/storage/data_stores/main/event_federation.py [hunk 3]
```diff
@@ -61,6 +62,28 @@ def get_auth_chain_ids(self, event_ids, include_given=False):
...
+            txn.execute(sql, (list(event_ids),))
+
+            results = set(event_id for event_id, in txn)
+
+            if include_given:
+                results.update(event_ids)
+
+            return list(results)
+
+        # Database doesn't necessarily support recursive CTE, so we falll
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/event_federation.py]:
```suggestion
        # Database doesn't necessarily support recursive CTE, so we fall
```

**Pull request merged**